JS: refactor most library models away from AST nodes

erik-krogh opened the pull request:
Initially I started with refactoring the HTTP models to use dataflow nodes.  
That was done surprisingly quick, so I continued with more library models.  

I've made deprecated aliases where a renaming made sense, and otherwise I just changed the type. 

The `flowsTo` predicates in the HTTP models were confusing when starting to use `DataFlow::Node`. (I initially made mistakes because of that). So I've deprecated those `flowsTo` predicates. 

## Evaluation

[An evaluation looks OK](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/httpNode__nightly-old__code-scanning__3/reports). But there are some new results and some lost results.  

### New results:  
All appear to be TPs, and they are all from recognizing destructuring assignments as property reads, which allows us to recognize more sources.   
Two of the results appeared from refactoring the HTTP models (the `js/path-injection` results).  
And two other appeared from refactoring the `SensitiveExpr` class (the `js/insufficient-password-hash` resutls).  

### Lost results:  

Both lost results are a consequence of a `DataFlow::PropWrite` having no result for `getRhs()` when the AST looks like:  
```JavaScript
obj.prop += value
```   
One appears to be a TP, the other appears to be an FP.  

I should do some followup work to investigate if we can have a sensible result for `getRhs()` in the above example.  (After this PR has merged).

erik-krogh commented on javascript/ql/lib/semmle/javascript/dataflow/TaintTracking.qll:
```diff
@@ -580,16 +576,16 @@ module TaintTracking {
-        pred = DataFlow::valueNode(assgn.getRhs()) and
+      exists(DataFlow::PropWrite assgn, DataFlow::SourceNode obj |
+        not exists(assgn.getPropertyName()) and
+        not assgn.getWriteNode() instanceof Property and // not a write inside an object literal
```
This line was added to preserve the behavior from the old predicate (new results appeared otherwise).  
I'm wondering if this line should be deleted?

annarailton commented:
Not qualified to sign off these changes to the `codeql-ml` code - will need to wait until @henrymercer is back on 6th April

asgerf commented on javascript/ql/lib/change-notes/2022-04-04-dataflow-models.md:
```diff
@@ -0,0 +1,5 @@
+category: minorAnalysis
+---
+* Many library models have been rewritten to use dataflow nodes instead of the AST.
+  These changes include some renamings where the old name is still available as a deprecated feature.
```
There are _some_ breaking changes, right? I think it's ok for a change like this, where it's practically unavoidable.

But if there are breaking changes, I think we should call out that some breaking changes have occurred, but in most cases, the AST-based predicates and classes are simply deprecated.

erik-krogh replied:
> There are some breaking changes, right? 

Definitely yes, I hit a decent amount of compile-errors while doing this refactoring.   
I've changed the category to `breaking`, and reworded the change-note a bit.

asgerf reviewed (commented):
What an incredible marathon of a rewrite! 👏 🎉 

I wasn't sure we'd ever get rid of this much legacy, but kudos for powering through here! 💪
I'm very happy to see this merged.

> Both lost results are a consequence of a DataFlow::PropWrite having no result for getRhs() when the AST looks like: `obj.prop += value`

I'm worried about losing `.innerHTML += x` as a sink, even temporarily.

Could you reduce the risk here, by making `DomSink` specifically add this back, perhaps via a helper predicate in `DomPropertyWrite`?

erik-krogh commented on javascript/ql/lib/semmle/javascript/frameworks/Express.qll:
```diff
@@ -901,9 +900,9 @@ module Express {
+     * This predicate is now from `DataFlow::SourceNode`.
      */
-    deprecated predicate flowsTo(Expr sink) { this.ref().flowsToExpr(sink) }
+    deprecated override predicate flowsTo(DataFlow::Node n) { super.flowsTo(n) }
```
It can be removed, but if you do that then previous uses of `RouterDefinition::flowsTo` would suddenly point to `SourceNode::flowsTo` without a deprecation warning, which would be a major change in behavior.  

That gotcha happened to me during this rewrite, and it caused weird behavior changes that took me a while to debug.   
So I prefer keeping this deprecation.

erik-krogh commented:
> I'm worried about losing .innerHTML += x as a sink, even temporarily.
> Could you reduce the risk here, by making DomSink specifically add this back, perhaps via a helper predicate in DomPropertyWrite? 

:+1: 
`DomPropertyWrite` already defined it's own `getRhs()` predicate, so I just added another case there.  

---  

I think I've addressed all your review comments.

asgerf commented:
Thanks. Would you mind running a final evaluation? It's a large change and we should reduce the risk as much as possible.

henrymercer reviewed (dismissed):
ATM 👍

erik-krogh commented:
> Thanks. Would you mind running a final evaluation? It's a large change and we should reduce the risk as much as possible.

[A new evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/httpNode-eval__default__code-scanning/reports) shows a small performance regression.  
But I haven't been able to nail down anything that's causing the regression.  
The two new results in the evaluation are both from destructuring assignment being seen as property reads.

erik-krogh commented:
I expanded the change-note to include which classes have changed type.  
I created it using a one-off QL-for-QL query (I had get both versions into the same DB).  

Here is the raw data from the query: https://docs.google.com/spreadsheets/d/1VkOGUofigegJYJ9un-qRczCBiEeNDEcTUsmFim90gRc/edit?usp=sharing

erik-krogh commented:
I ran the explicit-this patch to fix the alerts reported, and I fixed an instance of `ql/acronyms-should-be-pascal-case` in one of the new classes.   
So now there's no new QL-for-QL warnings from this PR.   

--- 

@asgerf I mentioned the classes with changed base-types in [the change-note](https://github.com/github/codeql/pull/8604/files#diff-e5cfab7a43a447ba2d6917edaff6e0488550c887dc0ddd29baa0d0306662ded1).  
I excluded deprecated classes and classes from Customizations files.  
So you think that works?

erik-krogh commented:
[New evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/httpNode__nightly-old__code-scanning__4/reports) looks similar to the previous evaluation.   
There is still a slight performance regression in the same benchmarks, but still no clear cause.   
It might just be from there being more rows to process.   
I still think we can merge it.

calumgrant commented:
Merge after Universe release.

erik-krogh commented:
The JavaScript language tests are broken, but that's unrelated.  
I got an internal PR that fixes it (see the backref).

erik-krogh commented:
[A new evaluation looks good](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-8604-4e14177__nightly-old__code-scanning__1/reports).  

The new results are due to recognizing object destructuring as property reads.

asgerf reviewed (approved):
👍
